kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/WaitForConditionWatcher.java

```diff
@@ -0,0 +1,62 @@
+package io.fabric8.kubernetes.client.dsl.base;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watcher;
+
+public class WaitForConditionWatcher<T extends HasMetadata> implements Watcher<T> {
+
+  private static final Logger log = LoggerFactory.getLogger(WaitForConditionWatcher.class);
+
+  private final Predicate<T> condition;
+  private final CompletableFuture<T> future;
+  private volatile boolean closed;
+
+  public WaitForConditionWatcher(Predicate<T> condition) {
+    this.condition = condition;
+    this.future = new CompletableFuture<>();
+  }
+
+  public CompletableFuture<T> getFuture() {
+    return future;
+  }
+
+  public boolean isClosed() {
+    return closed;
+  }
+
+  @Override
+  public void eventReceived(Action action, T resource) {
+    switch (action) {
+      case ADDED:
+      case MODIFIED:
+        if (condition.test(resource)) {
+          future.complete(resource);
+          break;
+        }
+      case DELETED:
+        if (condition.test(null)) {
+          future.complete(null);
+          break;
+        }
+      case ERROR:
+        log.debug("ERROR event for " + resource.getMetadata().getName());
```

Comment:
I wonder if we should complete here? its not even clear what this event would represent (is it an error on the resource? an error in the watcher stream?) but seems like we could get stuck here

Reply:
I was having a hard time figuring out what could come through on the ERROR. I noticed some actual error handling which delegated through onClose, so figured these were transient errors. However, lacking docs, I finally found how the native go client handles this. Will replicate